triton-vm/src/domain.rs

```diff
@@ -1,96 +1,100 @@
 use std::marker::PhantomData;
-use std::ops::MulAssign;
+use std::ops::{Mul, MulAssign};
 
 use twenty_first::shared_math::b_field_element::BFieldElement;
 use twenty_first::shared_math::polynomial::Polynomial;
 use twenty_first::shared_math::traits::{FiniteField, ModPowU32};
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct FriDomain<FF> {
+pub struct Domain<FF> {
     pub offset: BFieldElement,
-    pub omega: BFieldElement,
+    pub generator: BFieldElement,
     pub length: usize,
-    _field: PhantomData<FF>,
+    _finite_field: PhantomData<FF>,
 }
 
-impl<FF> FriDomain<FF>
+impl<FF> Domain<FF>
 where
-    FF: FiniteField + From<BFieldElement> + MulAssign<BFieldElement>,
+    FF: FiniteField
+        + From<BFieldElement>
+        + Mul<BFieldElement, Output = FF>
+        + MulAssign<BFieldElement>,
 {
-    pub fn new(offset: BFieldElement, omega: BFieldElement, length: usize) -> Self {
-        let _field = PhantomData;
+    pub fn new(offset: BFieldElement, generator: BFieldElement, length: usize) -> Self {
         Self {
             offset,
-            omega,
+            generator,
             length,
-            _field,
+            _finite_field: PhantomData,
         }
     }
 
-    pub fn evaluate(&self, polynomial: &Polynomial<FF>) -> Vec<FF> {
-        polynomial.fast_coset_evaluate(&self.offset, self.omega, self.length)
+    pub fn evaluate<GF>(&self, polynomial: &Polynomial<GF>) -> Vec<GF>
+    where
+        GF: FiniteField + From<BFieldElement> + MulAssign<BFieldElement>,
+    {
+        polynomial.fast_coset_evaluate(&self.offset, self.generator, self.length)
     }
 
-    pub fn interpolate(&self, values: &[FF]) -> Polynomial<FF> {
-        Polynomial::<FF>::fast_coset_interpolate(&self.offset, self.omega, values)
+    pub fn interpolate<GF>(&self, values: &[GF]) -> Polynomial<GF>
+    where
+        GF: FiniteField + From<BFieldElement> + MulAssign<BFieldElement>,
+    {
+        Polynomial::<GF>::fast_coset_interpolate(&self.offset, self.generator, values)
     }
```

Comment:
I'm happy to change the name. However, it can't be `FF`, because `GF` is actually the _second_ generic type parameter here. `FF`'s trait bounds are defined by the struct, `GF`'s trait bounds are defined by the two methods `evaluate()` and `interpolate()`.
The trait bounds for `GF` and `FF` are (or at least can be) identical. If there is some way to express this, I'm happy to learn it. If there isn't, would `FF2` be a better name than `GF`, since it more clearly indicates that the two generic types can be distinct?

Reply:
Perhaps I should explain the reason for having two generic types. My intention is to de-couple a polynomial's domain from its co-domain in order to use the most-fitting finite field for either.

Illustrating the mentioned de-coupling: polynomial $f \in \mathbb{Z}[x]$ can, without problem, be interpreted as function $f: \mathbb{Q} \rightarrow \mathbb{Q}, x \mapsto f(x)$; evaluating $f = 2\cdot x^3 - 5\cdot x$ on $x = \frac{2}{3}$ works like a charm.

However, before introducing the second generic type `GF` above, no such separation was possible in our code. This lead to entire domains being lifted from the B-field to the X-field, and consequently spending many CPU cycles multiplying zeros.

Reply:
> […] would `FF2` be a better name than `GF` […]

Dropping the generic type parameter `FF` from (what is currently named) `Domain` frees up that name to be used in the methods `evaluate()` and `interpolate()`, making this question superfluous.

Reply:
I'll leave this as is for the time being, with the definite intention to clear this up in #127.